File: [K]graph/[K]graph-matrix/54-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        rows, cols = len(matrix), len(matrix[0])
        visited = set()
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        r, c, d = 0, 0, 0
        res = []
        while len(res) < rows * cols:
            res.append(matrix[r][c])
            visited.add((r, c))
            next_r, next_c = r + directions[d][0], c + directions[d][1]
            if next_r not in range(rows) or next_c not in range(cols) or (next_r, next_c) in visited:
                d += 1
                d %= len(directions)
                next_r, next_c = r + directions[d][0], c + directions[d][1]
            r, c = next_r, next_c
        return res
    
# time O(RC)
# space O(RC)
# using graph and matrix and hashset

File: [K]graph/[K]graph-matrix/54-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int d = 0;
        int r = 0;
        int c = 0;
        int[][] directions = new int[][]{{0, 1}, {1, 0}, {0, - 1}, {- 1, 0}};
        HashSet<Integer> visited = new HashSet<>();
        List<Integer> res = new ArrayList<>();
        while (res.size() < rows * cols) {
            res.add(matrix[r][c]);
            visited.add(r * cols + c);
            int nextR = r + directions[d][0];
            int nextC = c + directions[d][1];
            if (0 > nextR || nextR >= rows || 0 > nextC || nextC >= cols || visited.contains(nextR * cols + nextC)) {
                d = (d + 1) % directions.length;
                nextR = r + directions[d][0];
                nextC = c + directions[d][1];
            }
            r = nextR;
            c = nextC;
        }
        return res;
    }
}

// time O(RC)
// space O(RC)
// using graph and matrix and hashset